<center> <h2> DS 3000 - Fall 2019</h2> </center>
<center> <h3> DS Report </h3> </center>


<center> <h1> Northeastern Transportation Analysis</h1> </center>
<center><h4>Benjamin Jacobson, Divit Koradia, Kyle Chu</h4></center>


<hr style="height:2px; border:none; color:black; background-color:black;">

#### Executive Summary:

Add your summary here (100-150 words)

Provide a brief summary of your project. After reading this executive summary, your readers should have a rough understanding of what you did in this project. You can think of this summary in terms of the four sections of the report and write 1-2 sentences describing each section.



 Have you ever wondered what is the best way to get from Mission Hill to Northeastern campus and vice versa? When you’re in a rush calculating the time and cost of different transportation types can be tedious. Our goal is to find an answer to the best transportation type in terms of speed and cost. We will be looking at MBTA, Uber, and Bluebikes. We will be comparing the average time and cost it takes people to make it from Mission Hill to Northeastern and vice versa. Our datasets are pulled from the different transportation types websites. The data includes the time it takes to complete trips, so for Uber the time from pickup to drop off, Bluebikes it’s time for a rider to go from one station to another, and for MBTA Roxbury Crossing to Ruggles and Brigham Circle to Northeastern. Bluebikes and Uber also have the cost of each trip. MBTA and Bluebikes also have the option for a fixed cost ride structure where you can pay one price for the whole trip or use a monthly or yearly pass. We plan to measure from a point on campus to a point on Mission Hill. We will take into account average / estimated walking speeds from a station to the point.


<hr style="height:2px; border:none; color:black; background-color:black;">

## Outline
1. <a href='#1'>INTRODUCTION</a>
2. <a href='#2'>METHOD</a>
3. <a href='#3'>RESULTS</a>
4. <a href='#4'>DISCUSSION</a>

<a id="1"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 1. INTRODUCTION

In this section, orient your readers to your project. You've already written some of these in previous deliverables. Based on your final analysis, revise your problem statement and write a concise introduction section. This section should touch upon the following points, but should be written in full paragraphs. Your writing should incorporate all of these points (and more if you like) in a coherent way. Remember that you are trying to convince your readers that this is an important problem to tackle. 

Problem Statement
* Describe the problem you would like to tackle.   
* What is the topic of your project? 
* What do you want to learn about it?    
    - We want to find the best transportation method between Northeastern and Mission Hill. Lots of students live off campus near mission hill, so we'd like to help them get to campus more effectively. 
    - Transportation between two points, based on personal preferences. 
    - Depending on the person, which mode of transportation would be the most effective for them to get to campus. 


Significance of the Problem
* Why is it important to tackle this problem in your project?
* In what ways could the insights from this project be useful?
* Has there been previous work on your topic? Do some research into your topic. Cite your sources appropriately. You can use the numbered reference format or APA (if you are more comfortable with it).

     - It's important for us to find the best mode of transportation for each individual person as Northeastern has a very large demographic of students living/commuting to and from Mission Hill. Based on how prospective users value time, money, distance, and wait-time, we'd like to show which mode of transportation would be best fit. There hasn't been previous work on our topic. The most similar reasearch has been in estimating prices in Boston, but not choosing optimal modes of transportation. 
     - 

Questions/Hypothesis
* End this section with a list of questions and hypotheses
* You should tie these questions/hypotheses to the problem statement and its significance
    * e.g. Given the aforementioned problem and its importance, we set out to tackle the following questions:
    
    
    Which mode of transportation is best:
    Which mode of transportation has the lowest average wait time for rides?
    Which mode of transportation has the lowest average trip time from point A to point B?
    Which mode of transportation has the least amount of walking on average?
    Which mode of transportation has the lowest average cost?

    
    H0: MBTA >= Uber and Bluebikes
    H1: MBTA < Uber and Bluebikes
    The average cost for a trip will be lowest for MBTA and highest for Uber(Cost / Distance, Cost / Time) 

    H0: Bluebikes >= MBTA and Bluebikes
    H1: Bluebikes < MBTA and Bluebikes
    Average total commute time will be lowest for Bluebike and highest for MBTA

    H0: Uber >= MBTA and Bluebikes
    H1: Uber < MBTA and Bluebikes
    Average walking time will be lowest for Uber and highest for MBTA

    H0: Bluebikes >= MBTA and Uber
    H1: Bluebikes < MBTA and Uber
    Average wait time will be lowest for Bluebikes and highest for MBTA


**Requirement:**
* You should have at least one question tapping into the comparison of various machine learning algorithms in predicting your target variable from your features variables.

<a id="2"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 2. METHOD

### 2.1. Data Acquisition

* Describe where you obtained your data. Provide a link to the original source. 
* If you scraped your data, include your code as a script file.
* Your data should be stored in an online repository (e.g., GitHub) and your code should retrieve your data from that online resource. You can read csv files from the Web in the same way that you read files from local drive.
* Describe the dataset and variables. What do variables represent?


In [ ]:
https://www.ctps.org/apps/mbtasurvey2018/#navButton
    
    
https://www.bluebikes.com/system-data
    
https://movement.uber.com/explore/boston/travel-times/query?si=302&ti=1118&ag=censustracts&dt[tpb]=ALL_DAY&dt[wd;]=1,2,3,4,5,6,7&dt[dr][sd]=2019-06-01&dt[dr][ed]=2019-06-30&cd=&sa;=-71.0908821,42.3391989&sdn=Northeastern%20University%20School%20of%20Law,%20416%20Huntington%20Ave,%20Boston,%20MA&lng.=-71.2121505&lat.=42.3599474&z.=11.49&ta;=-71.0993596,42.3330516&tdn=Mission%20Hill%20Playground,%20Boston,%20MA&lang=en-US
#NEU to Mission Hill




### 2.2. Variables
* If you are testing hypotheses, what are your IVs and DVs?
* For your predictive models, what are your features and target variables?


### 2.3. Data Analysis
* Specifically describe your predictive model. What outcome variable are you going to predict from what feature variables?
* Describe whether this is a supervised or unsupervised learning problem. Also identify the sub-category of the learning task (e.g. classification).
* What machine learning algorithms are you going to use? Why?

<a id="3"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 3. RESULTS

### 3.1. Data Wrangling
* Perform simple data cleaning (delete extra columns, deal with NA values, etc.)
* Perform data wrangling to extract your features and target values (e.g., grouping your dataframe by columns, applying functions to format dataframes, etc.)
* Preprocess your variables (e.g., scaling/transforming feature variables to normalize them)
* Feature extraction (dummy variables, new features from existing features, etc.)
* Use one feature selection technique to select a subset of your original features


### 3.2. Data Exploration
* Generate appropriate data visualizations for your key variables identified in the previous section
* You should have at least three visualizations (and at least two different visualization types)
* For each visualization provide an explanation regarding the variables involved and an interpretation of the graph.
* If you are using Plotly, insert your visualizations as images as well (upload the graph images to an online source, e.g. github, and link those in Jupyter Notebook)


In [5]:
import pandas as pd
import numpy as np

data_sets = {}

data_sets['bluebikes_data'] = pd.read_csv('./data_sets/Bluebikes_Data.csv')
data_sets['uber_ids'] = pd.read_csv('./data_sets/Uber_Data_IDs.csv')
data_sets['uber_mh_neu'] = pd.read_csv('./data_sets/Uber_Data_MH_NEU.csv')
data_sets['uber_neu_mh'] = pd.read_csv('./data_sets/Uber_Data_NEU_MH.csv')
data_sets['uber_neu_mh_monthly'] = pd.read_csv('./data_sets/Uber_Data_NEU_MH_Monthly.csv')

uber_tip_data = pd.read_csv('C:/Users/kchu1/Downloads/Transportation_Network_Providers_-_Trips.csv')

uber_tip_data.head()

D:\Anaconda\lib\site-packages\IPython\core\interactiveshell.py:3020: DtypeWarning: Columns (9,12) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Trip ID    Trip Start Timestamp  \
0  de3fd27d2229818c961bd74b3954d97c7bcc3664  11/01/2018 12:00:00 AM   
1  e03d12d43f4158884c2243ebb2c0234888086fa9  11/01/2018 12:00:00 AM   
2  e0b0aef3ed9d28edd30500bc08f8bee3fe9348db  11/01/2018 12:00:00 AM   
3  e0b1a906bbe625662cf40770d7ec43cb1aa311bb  11/01/2018 12:00:00 AM   
4  e0d82541bfc9de06172a45643d32330363010d1b  11/01/2018 12:00:00 AM   

       Trip End Timestamp Trip Seconds  Trip Miles  Pickup Census Tract  \
0  11/01/2018 12:15:00 AM          374    1.782258                  NaN   
1  11/01/2018 12:15:00 AM          408    2.150144                  NaN   
2  11/01/2018 12:15:00 AM          504    3.389895         1.703183e+10   
3  11/01/2018 12:15:00 AM          934    5.518289                  NaN   
4  11/01/2018 12:15:00 AM          737    3.834838         1.703108e+10   

   Dropoff Census Tract  Pickup Community Area  Dropoff Community Area Fare  \
0                   NaN                   58.0                    66.0  7.5   
1                   NaN                   31.0                    59.0    5   
2          1.703184e+10                   28.0                    33.0  7.5   
3                   NaN                    8.0                    35.0   10   
4          1.703107e+10                    8.0                     7.0  7.5   

                   ...                   Additional Charges  Trip Total  \
0                  ...                                  0.0         7.5   
1                  ...                                  0.0           5   
2                  ...                                  2.5          10   
3                  ...                                  0.0          10   
4                  ...                                  0.0         8.5   

  Shared Trip Authorized  Trips Pooled  Pickup Centroid Latitude  \
0                   True             1                 41.817366   
1                   True             1                 41.850266   
2                  False             1                 41.879067   
3                   True             1                 41.899602   
4                   True             1                 41.892042   

   Pickup Centroid Longitude              Pickup Centroid Location  \
0                 -87.698861  POINT (-87.6988607973 41.8173662079)   
1                 -87.667569   POINT (-87.667569312 41.8502663663)   
2                 -87.657005   POINT (-87.657005027 41.8790669938)   
3                 -87.633308   POINT (-87.6333080367 41.899602111)   
4                 -87.631864  POINT (-87.6318639497 41.8920421365)   

  Dropoff Centroid Latitude  Dropoff Centroid Longitude  \
0                 41.771849                  -87.695666   
1                 41.829922                  -87.672503   
2                 41.849247                  -87.624135   
3                 41.835118                  -87.618678   
4                 41.930579                  -87.642206   

              Dropoff Centroid Location  
0   POINT (-87.695666342 41.7718485152)  
1  POINT (-87.6725026464 41.8299223039)  
2  POINT (-87.6241352979 41.8492467545)  
3  POINT (-87.6186777673 41.8351179863)  
4  POINT (-87.6422063127 41.9305785697)  

[5 rows x 21 columns]

In [7]:
uber_tip_data_random = uber_tip_data.sample(n=1000000)
uber_tip_data_random.head()


Trip ID    Trip Start Timestamp  \
29907868  7c239ad0a936e1b01d3f3e898b455f7fe8bb50fc  02/12/2019 07:15:00 PM   
39759310  589e80cc8fdbd99a089afe483954e8eb546c8d96  03/14/2019 10:15:00 PM   
43596650  8134aa016415900b547c6cf806d7f3f9a28fb892  03/26/2019 01:30:00 PM   
4573589   99093b3ecd45e2f76bf27c14ef27da8a3cca23d6  11/15/2018 11:00:00 PM   
28449485  036d89e174aad323c06864bf7842fd9eb180706d  02/08/2019 08:30:00 AM   

              Trip End Timestamp Trip Seconds  Trip Miles  \
29907868  02/12/2019 07:15:00 PM          248    0.746336   
39759310  03/14/2019 10:30:00 PM          900    5.371249   
43596650  03/26/2019 01:45:00 PM        1,009    5.669338   
4573589   11/15/2018 11:15:00 PM          733    6.879594   
28449485  02/08/2019 08:45:00 AM          675    3.251938   

          Pickup Census Tract  Dropoff Census Tract  Pickup Community Area  \
29907868         1.703103e+10          1.703103e+10                    3.0   
39759310         1.703108e+10          1.703122e+10                    8.0   
43596650         1.703133e+10          1.703124e+10                   33.0   
4573589          1.703198e+10                   NaN                   76.0   
28449485         1.703110e+10          1.703181e+10                   10.0   

          Dropoff Community Area Fare                  ...                   \
29907868                     3.0  2.5                  ...                    
39759310                    22.0    5                  ...                    
43596650                    24.0    5                  ...                    
4573589                      NaN   10                  ...                    
28449485                     NaN   10                  ...                    

          Additional Charges  Trip Total Shared Trip Authorized  Trips Pooled  \
29907868                2.55        5.05                  False             1   
39759310                0.72        5.72                   True             2   
43596650                2.55        7.55                   True             4   
4573589                 9.52       19.52                  False             1   
28449485                0.00          10                   True             1   

          Pickup Centroid Latitude  Pickup Centroid Longitude  \
29907868                 41.965446                 -87.663196   
39759310                 41.899156                 -87.626211   
43596650                 41.861281                 -87.630580   
4573589                  41.979071                 -87.903040   
28449485                 41.995603                 -87.794269   

                      Pickup Centroid Location Dropoff Centroid Latitude  \
29907868  POINT (-87.6631958495 41.9654457842)                 41.972710   
39759310  POINT (-87.6262105324 41.8991556134)                 41.921273   
43596650  POINT (-87.6305800611 41.8612808467)                 41.906026   
4573589   POINT (-87.9030396611 41.9790708201)                       NaN   
28449485  POINT (-87.7942692959 41.9956031347)                       NaN   

          Dropoff Centroid Longitude             Dropoff Centroid Location  
29907868                  -87.657341  POINT (-87.6573411073 41.9727095466)  
39759310                  -87.685082  POINT (-87.6850821101 41.9212731053)  
43596650                  -87.675312   POINT (-87.6753116216 41.906025969)  
4573589                          NaN                                   NaN  
28449485                         NaN                                   NaN  

[5 rows x 21 columns]

In [9]:
uber_tip_data_random.to_csv('Uber_Data_Tips', encoding='utf-8', index=False)

In [11]:
data_sets['uber_tips'] = pd.read_csv('./Uber_Data_Tips.csv')
data_sets['uber_tips'].shape

(1000000, 14)

### 3.3. Model Construction
* If you proposed hypotheses, conduct your hypothesis tests
* For your machine learning question(s), split data into training, validation, and testing sets (or use cross-validation)
* Apply machine learning algorithms (apply at least three algorithms)
* Train your algorithms

### 3.4. Model Evaluation
* Evaluate the performance of your algorithms on appropriate evaluation metrics, using your validation set
* Interpret your results from multiple models (and hypothesis tests, if any)

### 3.5. Model Optimization
* Tune your models using appropriate hyperparameters
* Explain why you are doing this (e.g., to avoid overfitting, etc.)

### 3.6. Model Testing
* Test your tuned algorithms using your testing set

<a id="4"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 4. DISCUSSION
* Provide a summary of the steps you took to analyze your data and test your predictive model
* Intepret your findings from 3.4., 3.5, and 3.6
    * Which algorithms did you compare?
    * Which algorithm(s) revealed best performance?
    * Which algorithm(s) should be used for your predictive model?
* If you tested hypotheses, interpret the results. What does it mean to have significant/non-significant differences with regards to your data?


* End this section with a conclusion paragraph containing some pointers for future work 
    *(e.g., get more data, perform another analysis, etc.)

<a id="5"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

### CONTRIBUTIONS
* Describe each team member's contributions to the report (who did what in each section)
* Remember this is a team effort!
* Each member of your team will provide peer evaluation of other team members. Your final grade on the project will be based on those peer evaluations. 